### ISSUE TITLE
The promise interface of adding .output to the class is strange and breaks expectations

### ISSUE THREAD

**Issue by vincentwoo** [apocas/dockerode#486]
I've been translating my use of dockerode from callbacks to Promises. I hit a very mysterious bug recently. Previously, I had code that looked like:

```js
.then(exec =>
  Promise.fromCallback(cb =>
    exec.start(
      {
        Detach: false,
        stderr: true,
        stdin: true,
        stdout: true,
        stream: true,
        Tty: true
      },
      cb // Returns a stream.
    )
  )
)
.then(stream => doSomethingWithStream(stream))
```

Now that dockerode supports promises on its own, I thought I could do:

```js
.then(exec =>
  exec.start({
    Detach: false,
    stderr: true,
    stdin: true,
    stdout: true,
    stream: true,
    Tty: true
  })
)
.then(stream => doSomethingWithStream(stream))
```

To my great surprise, I found that in the final `then` callback, the value of `stream` was not a stream, but a copy of exec. I believe this is because of dockerode's policy of not resolving promises with what the callback would have received, but instead adding a special `output` key to the object in question and resolving the promise with the object: https://github.com/apocas/dockerode/blob/311b0e1c6876e6e4b6c5d7520523995e4a667ac9/lib/exec.js#L50

I think that this is a very misleading thing to have in this library, and would suggest removing it as soon as possible, or at least documenting it thoroughly. I also strongly recommend consolidating the promise/callback selection logic into one utility method. Right now, the block of code that checks for the presence of a callback is repeated in every dockerode method.

**Comment by apocas** [comment]
This was done long time ago, can't remember the details.

But I think it was done in order to make chaining easier.

**Comment by vincentwoo** [comment]
Do you agree this should be changed to be more consistent with the callback interface?

**Comment by apocas** [comment]
Sure. We could prepare this for the next major release :)

**Comment by vincentwoo** [comment]
I think a good start would be to replace all the method conclusions that look like:

```js
if(args.callback === undefined) {
  return new this.modem.Promise(function(resolve, reject) {
    self.modem.dial(optsf, function(err, data) {
      if (err) {
        return reject(err);
      }
      self.output = data;
      resolve(self);
    });
  });
} else {
  this.modem.dial(optsf, function(err, data) {
    if (err) return args.callback(err, data);
    args.callback(err, data);
  });
}
```

with a call to something like:

```js
return utils.wrappedDial(this.modem, optsf, callback)
```

that is implemented like:

```js
function wrappedDial(modem, optsf, callback) {
  if (callback === undefined) {
    return new modem.Promise(function(resolve, reject) {
      modem.dial(optsf, function(err, data) {
        if (err) {
          return reject(err);
        }
        resolve(data);
      });
    });
  } else {
    modem.dial(optsf, callback);
  }
}
```

My brief inspection seemed to show that every method ends with:

```js
  this.modem.dial(optsf, function(err, data) {
    if (err) return args.callback(err, data);
    args.callback(err, data);
  });
```

so pulling that into a utility method seems like a safe bet.

**Comment by apocas** [comment]
Finally got some time to prep the next major release and do some refactoring. This was one of them.

**Comment by apocas** [comment]
It's in the master branch, going to be published in v3.0.0
